packages/grafana-ui/src/components/DataSourceDrawer/DataSourceDrawer.tsx

```diff
@@ -0,0 +1,199 @@
+import { css } from '@emotion/css';
+import React, { useCallback, useState } from 'react';
+
+import { DataSourceInstanceSettings, DataSourceJsonData, DataSourceRef, GrafanaTheme2 } from '@grafana/data';
+
+import { useStyles2 } from '../../themes';
+import { Button } from '../Button';
+import { Card } from '../Card/Card';
+import CustomScrollbar from '../CustomScrollbar/CustomScrollbar';
+import { Drawer } from '../Drawer/Drawer';
+import { FileDropzone, FileDropzoneDefaultChildren } from '../FileDropzone';
+import { Input } from '../Input/Input';
+import { ModalsController } from '../Modal/ModalsContext';
+import { PluginSignatureBadge } from '../PluginSignatureBadge/PluginSignatureBadge';
+import { Tag } from '../Tags/Tag';
+
+import { PickerContentProps, DataSourceDrawerProps } from './types';
+
+export function DataSourceDrawer(props: DataSourceDrawerProps) {
+  const { current, onChange } = props;
+  const styles = useStyles2(getStyles);
+
+  return (
+    <ModalsController>
+      {({ showModal, hideModal }) => (
+        <Button
+          className={styles.picker}
+          onClick={() => {
+            showModal(PickerContent, {
+              ...props,
+              onDismiss: hideModal,
+              onChange: (ds) => {
+                onChange(ds);
+                hideModal();
+              },
+            });
+          }}
+        >
+          <DataSourceDisplay ds={current}></DataSourceDisplay>
+        </Button>
+      )}
+    </ModalsController>
+  );
+}
+
+function DataSourceDisplay(props: {
+  ds: DataSourceInstanceSettings<DataSourceJsonData> | string | DataSourceRef | null | undefined;
+}) {
+  const { ds: ds } = props;
+  const styles = useStyles2(getStyles);
+
+  if (!ds) {
+    return <span>Unknown</span>;
+  }
+
+  if (typeof ds === 'string') {
+    return <span>${ds} - not found</span>;
+  }
+
+  if ('name' in ds) {
+    return (
+      <>
+        <img className={styles.pickerDSLogo} alt={`${ds.meta.name} logo`} src={ds.meta.info.logos.small}></img>
+        <span>{ds.name}</span>
+      </>
+    );
+  }
+
+  return <span>{ds.uid} - not found</span>;
+}
+
+function PickerContent(props: PickerContentProps) {
```

Comment:
Given how complicated this will likely get -- and the future interation with internal APIs, maybe it would be better to have the implementation in core while keeping access for plugins.  Similar to what we do in:

https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/components/PanelRenderer.tsx#L43

Reply:
Yes! Thank you! I was looking for a way to keep this in core, but deferred when I couldn't find a way forward.